3.0 fix query regression

lorenzo opened the pull request:
Fixes https://github.com/cakephp/cakephp/issues/5721

lorenzo committed:
Initial fix for conflict between contain and matching.

Currently it only works for MySQL

lorenzo committed:
Postgres and SQL Server require an order clause for subqueries having a distinct

lorenzo committed:
Now going back to the SELECT strategy so it can be fixed as well

lorenzo committed:
Fixed eagerloading with matching and contain involving belongsTo associations

lorenzo committed:
Adding reference to the issue that was fixed

lorenzo committed:
Merge remote-tracking branch 'origin/3.0' into 3.0-fix-query-regression

Conflicts:
	src/ORM/EagerLoader.php
	tests/TestCase/ORM/QueryRegressionTest.php

lorenzo committed:
An attempt to fix errors in SQL server

markstory commented on src/ORM/EagerLoader.php:
```diff
@@ -656,13 +658,7 @@ protected function _groupKeys($statement, $collectKeys)
-        foreach ($collectKeys as $parts) {
-            if ($parts[2] === true && isset($keys[$parts[0]])) {
-                $keys[$parts[0]] = array_unique($keys[$parts[0]]);
+                $keys[$nestKey][$parts[0]][implode(';', $collected)] = $collected;
```
What about dupes? How are those being removed now?

markstory commented:
A new failure in the sqlserver driver.

markstory committed:
Fix errors in SQLServer 2008

Because SQLServer 2008 doesn't have native pagination we don't want to
try and select the _cake_row_num_ field which is an expression.
Generally if the query is ordered by a subquery or expression object we
probably don't want to select that.

Refs #5817

markstory commented:
I think I have the SQLServer test fixed, I guess we'll see.

lorenzo committed:
Using Group By instead of Distinct as it has less shortcomings

lorenzo committed:
Removing debug code
